policies/recipes/artifacts_test.go

```diff
@@ -0,0 +1,302 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package recipes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+
+	"cloud.google.com/go/storage"
+	osconfigpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha2"
+	"github.com/fsouza/fake-gcs-server/fakestorage"
+)
+
+func TestFetchArtifacts_GCS_happyCase(t *testing.T) {
+	server := fakestorage.NewServer([]fakestorage.Object{
+		{
+			BucketName: "some-bucket",
+			Name:       "some-object.txt",
+			Content:    []byte("inside the file"),
+		},
+	})
+	defer server.Stop()
+
+	getGCSClient = func(ctx context.Context) (*storage.Client, error) {
+		return server.Client(), nil
+	}
+
+	gcsartifact := osconfigpb.SoftwareRecipe_Artifact_Gcs{
+		Object:     "some-object.txt",
+		Bucket:     "some-bucket",
+		Generation: 1,
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Gcs_{Gcs: &gcsartifact},
+			AllowInsecure: true,
+		},
+	}
+	FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if !exists("/tmp/id1.txt") {
+		t.Errorf("expected fetched artifacts")
+	}
+
+	if err = os.Remove("/tmp/id1.txt"); err != nil {
+		t.Fatalf("could not clean up test data")
+	}
+
+	if art["id1"] != "/tmp/id1.txt" {
+		t.Errorf("expected artifact entry(%s), got(%s)\n", "/tmp/id1", art["id1"])
+	}
+}
+
+func TestFetchArtifacts_GCS_ReadError(t *testing.T) {
+	server := fakestorage.NewServer([]fakestorage.Object{
+		{
+			BucketName: "some-bucket",
+			Name:       "some-object.txt",
+			Content:    []byte("inside the file"),
+		},
+	})
+	defer server.Stop()
+
+	getGCSClient = func(ctx context.Context) (*storage.Client, error) {
+		return server.Client(), nil
+	}
+
+	gcsartifact := osconfigpb.SoftwareRecipe_Artifact_Gcs{
+		Object:     "some-other-object.txt",
+		Bucket:     "some-bucket",
+		Generation: 1,
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Gcs_{Gcs: &gcsartifact},
+			AllowInsecure: true,
+		},
+	}
+	FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpted artifact fetched")
+	}
+
+	if err == nil {
+		t.Errorf("Expected (reader error), got(%+v)\n", err)
+	}
+}
+
+func TestFetchArtifacts_GCS_ClientError(t *testing.T) {
```

Comment:
this function checks whether our 'if err != nil' line in FetchArtifacts actually returns. drop this function.

Reply:
since the client creation is refactored out; the test need not exist anymore. dropped this test